File: Android/app/src/main/java/com/hhp227/datemate/data/UserRepository.kt
package com.hhp227.datemate.data

import com.google.firebase.auth.FirebaseUser
import com.hhp227.datemate.common.Resource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart

class UserRepository private constructor(
    private val userRemoteDataSource: UserRemoteDataSource
) {
    val loginStateFlow = userRemoteDataSource.userStateFlow
        .map { if (it != null) LoginState.Login else LoginState.Logout }
        .onStart { emit(LoginState.Loading) }

    fun getSignInResultStream(email: String, password: String): Flow<Resource<FirebaseUser>> {
        return userRemoteDataSource.signIn(email, password)
    }

    fun getSignOutResultStream(): Flow<Resource<Boolean>> {
        return userRemoteDataSource.signOut()
    }

    enum class LoginState {
        Login, Logout, Loading
    }

    companion object {
        @Volatile private var instance: UserRepository? = null

        fun getInstance(userRemoteDataSource: UserRemoteDataSource) =
            instance ?: synchronized(this) {
                instance ?: UserRepository(userRemoteDataSource).also { instance = it }
            }
    }
}

File: iOS/DateMate/Data/UserRepository.swift
//
//  UserRepository.swift
//  DateMate
//
//  Created by 홍희표 on 2021/12/01.
//

import Foundation
import FirebaseAuth
import Combine

class UserRepository {
    private let userRemoteDataSource: UserRemoteDataSource
    
    private var cancellables = Set<AnyCancellable>()
    
    @Published var signInState: SignInState = .loading
    
    var signInStatePublisher: AnyPublisher<SignInState, Never> {
        $signInState.eraseToAnyPublisher()
    }
    
    func getSignInResultStream(email: String, password: String) -> AnyPublisher<Resource<User>, Never> {
        userRemoteDataSource.signIn(email: email, password: password)
    }

    func getSignOutResultStream() -> AnyPublisher<Resource<Bool>, Never> {
        userRemoteDataSource.signOut()
    }
    
    init(_ userRemoteDataSource: UserRemoteDataSource) {
        self.userRemoteDataSource = userRemoteDataSource
        
        userRemoteDataSource.userStatePublisher
            .map { $0 != nil ? SignInState.signIn : SignInState.signOut }
            .prepend(.loading)
            .receive(on: DispatchQueue.main)
            .assign(to: \.signInState, on: self)
            .store(in: &cancellables)
    }
    
    private static var instance: UserRepository? = nil
    
    static func getInstance(userRemoteDataSource: UserRemoteDataSource) -> UserRepository {
        if let instance = self.instance {
            return instance
        } else {
            let userRepository = UserRepository(userRemoteDataSource)
            self.instance = userRepository
            return userRepository
        }
    }
}

enum SignInState {
    case signIn
    case signOut
    case loading
}
